### DISCUSSION TITLE
LastSeen or Last Sent/Received

### DISCUSSION THREAD

**Discussion by b00f** [pactus-project/pactus#566] [General]
Currently, when we receive data from a peer, we update the [LastSeen](https://github.com/pactus-project/pactus/blob/f01370fe659c6f9e42cc2f116866562d22561501/sync/peerset/peer.go#L24) property.  

In Bitcoin, they have two properties per peer:  `lastsend` and `lastrecv`. Look [here](https://developer.bitcoin.org/reference/rpc/getpeerinfo.html)

Should we maintain a record of the last send/receive, or should we just keep the last seen property?

**Comment by b00f** [comment]
@amirvalhalla What is your opinion?

**Reply by kehiy** [reply]
why we need the lastsend property?
even the [ethereum](https://github.com/ethereum/go-ethereum/blob/6ca3ef9a7b7ebf3ed0004cfdca0c7a27a38c28e7/p2p/peer.go#L106) dosen't have this property

**Reply by amirvalhalla** [reply]
I think we don't need to  `lastsend` property and maybe it would be better to change the name of `lastseen` to  `lastrecv` it's more meaningful than `last seen`, I guess

@b00f

**Reply by amirvalhalla** [reply]
if `Ethereum` have these properties or not, it's related to `Ethereum` structure, yes we can get some help from `Ethereum` or other blockchain networks but we should consider our own system.

@kehiy

**Reply by kehiy** [reply]
@amirvalhalla
must of blockchain are follow `Ethereum` structure
but in general I think the pactus approach for this (`lastseen`) is the best way
Unlike Ethereum we have is so we can use it for something like GC
and Unlike bitcoin we don't have something unnecessary like last send!

**Reply by amirvalhalla** [reply]
@kehiy 

I agree with `lastseen` but the name of property not very well because peers will receive not seen in my opinion , it would be better to convert the name to `lastrecv` like `Bitcoin`

**Reply by kehiy** [reply]
@amirvalhalla 
sure, in our case `lastrecv` make more sense instead of `lastseen`
its more clear

**Reply by b00f** [reply]
As @amirvalhalla suggested, it is better to change `last seen` to `last received`. This property enables each node to determine whether a specific peer in the network is offline. This feature specifically allows us to implement a [garbage collector](https://github.com/pactus-project/pactus/issues/546), which removes offline nodes from the PeerSet.

I'm wondering, is there any situation in which we might be interested in knowing the last time we sent data to a peer (or `last sent`)?

Let me elaborate a bit:

We have two methods to send data: [Broadcast](https://github.com/pactus-project/pactus/blob/f01370fe659c6f9e42cc2f116866562d22561501/network/interface.go#L75) (Gossiping) and [Direct Send](https://github.com/pactus-project/pactus/blob/f01370fe659c6f9e42cc2f116866562d22561501/network/interface.go#L76) (Streaming): 

- In broadcast mode, we **gossip** data to all peers in the network, such as a [vote](https://github.com/pactus-project/pactus/blob/main/sync/bundle/message/vote.go) message.
- In direct send mode, we **stream** data directly to a specific node. There is a direct channel between two peers. Right now, we have only one scenario for streaming: `Block Request -> Block Response`. If a node is syncing itself, it may request data from a peer and in response it gets the missed blocks. 

Therefore, the last sent timestamp could may tell us when this node was last fully synced and didn't go offline and never missed any new blocks.

**Reply by amirvalhalla** [reply]
by this scenario , I Agree with `lastsent` property because we should know when a node send data for the last time.

**Reply by b00f** [reply]
Good news that only one place we need to change: [sendTo](https://github.com/pactus-project/pactus/blob/723463df60e39af3f53178881b99ba193d0180bb/sync/sync.go#L318C27-L318C27) function in sync package. 
Then we can add this property to [proto](https://github.com/pactus-project/pactus/blob/723463df60e39af3f53178881b99ba193d0180bb/www/grpc/proto/network.proto#L29) file and as well in [HTTP](https://github.com/pactus-project/pactus/blob/723463df60e39af3f53178881b99ba193d0180bb/www/http/blockchain.go#L43C22-L43C22)

**Reply by amirvalhalla** [reply]
@b00f 

okay , I'll refactor this section as soon as possible
